### DISCUSSION TITLE
Fl_GTK_Native_File_Chooser_Driver() automatically adds the "All Files" filter

### DISCUSSION THREAD

**Discussion by chezbel** [fltk/fltk#1106] [Development]
I'm using Ubuntu 24.04 (X11, not Wayland) and playing with Fl_Native_File_Chooser.

In Fl_GTK_Native_File_Chooser_Driver() when GTK_FILE_CHOOSER_ACTION_OPEN is specified, it automatically adds "All Files \t*" as the last filter item.

I think it would be better **not** to do this and leave it up to the app developer to include or not include "All Files" as a filter. Also, in the current implementation I couldn't see anyway to choose that option using filter_value(n) because it is outside the array I created with filter(). Specifying an **n** past the end of my array causes the first filter to be chosen and using -1 (unsurprisingly) causes a segfault.

I tried an experiment, which is in the attached chooser.patch, which just comments out the code that adds the "All Files \t*" filter. In my tests this works fine, although you might want to modify test/native-filechooser to now specify "All Files \t*"; I tried that manually and it works.

With my patch applied I tried this in my app and testing fileChooser->filter_value(n) with each value 0..4 works as expected.
```
fileChooser->filter("All Files \t*\n"
                    "Text|C++ \t*.{txt,c,cpp,cxx,h,hpp}\n"
                    "C++ \t*.{c,cpp,cxx,h,hpp}\n"
                    "Text \t*.{txt}\n"
                    "Perl \t*.pl\n"
                   );
```
If I use the above filter example with the unmodified FLTK source, then at runtime the list of filters will show "All Files" both at the top and bottom of the filter list; obviously this works, but it doesn't look as nice as it could.

If you think that using my patch might be annoying for existing apps, then perhaps we can have an option to specify not to add "All Files" automatically.

[chooser.PATCH](https://github.com/user-attachments/files/17614420/chooser.PATCH)

**Comment by Albrecht-S** [comment]
Hi @chezbel, thank you for your report.

I see your point but I'm currently too busy with other stuff (release 1.4.0) to reply right now with a **_final_** comment.

I think what we should do is:
- compare with other "native" filechoosers and the internal FLTK filechooser: what filters do they add?
- if other filechoosers don't add it, then we should remove it from the GTK filechooser as well (for consistency)
- we could add an option as you suggest.

The main point is IMHO that all file choosers behave similarly and - last, but not least - that program developers are free to add the filters they like.

**Reply by chezbel** [reply]
Hi Albrecht, thanks for your reply. I understand that you are busy with the 1.4.0 release and my issue is not urgent. I'll wait and see what you guys decide to do about this.

**Comment by chezbel** [comment]
Now that work is being done on 1.5, is this  item worth looking at again?

**Reply by Albrecht-S** [reply]
Yes, I agree, and as long as this discussion is "open" it won't be forgotten. However, my personal spare time is limited, I have other stuff on my to-do list, and I'm not very familiar with our different filechoosers.

I believe @ManoloFLTK is our "filechooser guy", and IIRC he developed the GTK filechooser. Manolo, could you maybe take a look at this issue? As I wrote before, if other filechoosers don't add the "All Files" filter, then it should be removed from the GTK chooser, otherwise we need to discuss a proper solution.

**Reply by ManoloFLTK** [reply]
I think that was done to reproduce what FLTK own file chooser does: it adds "All Files (*)" at the end of any user-defined pattern.

**Reply by Albrecht-S** [reply]
@ManoloFLTK Thanks for checking this. Then I think we should close this discussion because the behavior is intended and not GTK_Native_File_Chooser specific.

@chezbel would it be OK to close the issue w/o any changes?

**Reply by chezbel** [reply]
Sure, no problem. It's easy enough to patch locally whenever I update the FLTK source.

**Reply by Albrecht-S** [reply]
Thanks, closing ...
